### PR TITLE
added output_limiting_function parameter

### LINKED ISSUE #24694: Enable output based on function times
## Reason
This can force output times more efficiently for the user. 

## Design
Used piecewise base function domain to define sync_times in outputs. 

## Impact
Something new

### PR DESCRIPTION
**Pull request by joseph-silberman**
executed in output/outputWarehouse, and included relevant testing files closes #24694

### CODE REVIEW

**Comment by YaqiWang** [comment]
Maybe we can have a bool parameter `add_sync_times_with_timestepper` to inform the output system to grab the timestepper and insert output times based on the stepper? If the stepper is `FunctionDT`, we can retrieve the time knots. I am not sure what will happen if the time included in the output is not stepped by the time stepper.

**Comment by GiudGiud** [comment]
This seems fine but we're heading to a bloat if we just enable everyone to pull in times from wherever they want. 
I think something along the lines of what I propose in #24696 could satisfy everyone

@joseph-silberman what is your timeline for your needs here?

**Comment by jasondhales** [comment]
What we need is output at only the times defined by the Function.  The timesteps taken by the solver may be more frequent than these times.

These changes are needed for the next phase of Joseph's work, where he'll take the output at these specific times and compare it to archived or experimental data.

File: framework/src/outputs/Output.C [hunk 1]
```diff
@@ -156,6 +161,17 @@ Output::Output(const InputParameters & parameters)
     for (auto & me : add)
       _execute_on.push_back(me);
   }
+
+  if (isParamValid("output_limiting_function"))
+  {
+    const Function & olf = getFunction("output_limiting_function");
+    const PiecewiseBase * pwb_olf = dynamic_cast<const PiecewiseBase *>(&olf);
+    if (pwb_olf == nullptr)
+      mooseError("Function muse have a piecewise base!");
```
Comment (lindsayad) [inline framework/src/outputs/Output.C:170]: I think this code illustrates this design is flawed. It only works with specific types of functions. When I think of a `Function` I think of something that is conceptually continuous. So the idea of a continuous function dictating discrete output times doesn't seem right

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/src/outputs/Output.C (hunk 1 above)
Comment (jasondhales) [reply framework/src/outputs/Output.C:170]: This approach is the same as is used in IterationAdaptiveDT.  Due to the fact that IterationAdaptiveDT guarantees a solution step at the times listed in a piecewise linear function, we want to get output at exactly those times without needing to type the times one by one, which is the only option offered now.

Comment (lindsayad) [reply framework/src/outputs/Output.C:170]: The thing is that I think the limiting function in `IterationAdaptiveDT` should be changed to allow general `Functions`. You could write a ParsedFunction for what you want your max dt to be for a given time for example

Comment (jasondhales) [reply framework/src/outputs/Output.C:170]: Maybe there is some confusion about what these `Functions` are providing.  In our use, they are giving us the points in time when a solve must occur.  These times are the time values in the piecewise functions.  They are not specifying the dt.  The dt may be adjusted to hit the times from the functions, of course.

Comment (lindsayad) [reply framework/src/outputs/Output.C:170]: I find the function use in `IterationAdaptiveDT` confusing 😄 That class has become a monster and we are trying to make it less important or just do less with the new time step compositing possible with @MengnanLi91 and @GiudGiud's work. 

So I would prefer the design of `IterationAdaptiveDT` not be a template for other design decisions. I think that something like what @GiudGiud proposed in #24696 would be much better.

Comment (jasondhales) [reply framework/src/outputs/Output.C:170]: How long will @GiudGiud's solution take?  We can limp along with Joseph's fix local in a submodule, but that's not ideal.

Comment (GiudGiud) [reply framework/src/outputs/Output.C:170]: Not long I ll try my best

Comment (lindsayad) [reply framework/src/outputs/Output.C:170]: @jasondhales a short-term solution I could live with is adding this new parameter as a deprecated parameter to indicate to users that this is not something they should rely on existing in the future

**Review by GiudGiud — APPROVED** [review]

**Comment by GiudGiud** [comment]
Making progress on the Times system so we ll see

**Review by lindsayad — APPROVED** [review]

**Pull request merged**